### PR TITLE
feat: Proposal Authoring Improvements

### PR DESCRIPTION
**Pull request by ncomerci**
This PR closes [this issue](https://www.notion.so/dcl-dao/Proposal-Authoring-Improvements-164e1c5ff6554705981fce0d2f4247fa)

image
image
image
image
image
image
screencapture-pr375-test-decentraland-vote-proposal-2022-06-21-09_54_05

### CODE REVIEW

File: src/pages/index.tsx [hunk 1]
```diff
@@ -84,6 +87,9 @@ export default function IndexPage() {
     }
   }, [handlePageFilter, page, proposals])
 
+  const [user] = useAuthContext()
+  const [pendingCoathoring] = useCoauthoring(user, CoauthorStatus.PENDING)
```
Comment (1emu) [inline src/pages/index.tsx:91]: typo in pendingCoathoring

**Review by 1emu — APPROVED** [review]
Great work @ncomerci 👏 👏 👏

File: src/api/Governance.ts [hunk 2]
```diff
@@ -285,4 +287,25 @@ export class Governance extends GovernanceAPI {
...
+    )
+    return result.data
+  }
+
+  async getCoAuthorsByProposal(id: string, status?: CoauthorStatus) {
+    const result = await this.fetch<ApiResponse<CoauthorAttributes[]>>(`/coauthors/${id}${status ? `/${status}` : ''}`)
+    return result.data
+  }
+
+  async updateCoauthorStatus(proposalId: string, status: CoauthorStatus) {
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/api/Governance.ts:303]: Users can only update the status of their coauthoring requests, for which they need to be authenticated. The user's address together with the proposal ID form the key to be updated.

image

File: src/entities/Proposal/model.ts [hunk 3]
```diff
@@ -181,29 +183,29 @@ export default class ProposalModel extends Model<ProposalAttributes> {
...
+    FROM ${table(ProposalModel)} p
+        ${conditional(!!filter.subscribed, SQL`INNER JOIN ${table(SubscriptionModel)} s ON s."proposal_id" = p."id"`)} 
+        ${conditional(!!filter.coauthor, SQL`INNER JOIN ${table(CoauthorModel)} c ON c."proposal_id" = p."id"`)}
+      ${conditional(
+        !!filter.search,
+        SQL`, ts_rank_cd(p.textsearch, to_tsquery(${tsquery(filter.search || '')})) AS "rank"`
+      )}
+    WHERE "deleted" = FALSE 
+    ${conditional(!!filter.user && !filter.coauthor, SQL`AND p."user" = ${filter.user}`)} 
+    ${conditional(!!filter.coauthor, SQL`AND lower(c."coauthor_address") = lower(${filter.user})`)}
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/entities/Proposal/model.ts:203]: filter.coauthor is defined as boolean, the address is stored in user

image

File: src/theme.css [hunk 4]
```diff
@@ -135,25 +135,30 @@ body {
...
+  --background-linked-wearables: #ffbc5b;
   --background-popup: #5c5c6e;
 
-  --blue: #4888EE;
-  --gray: #A19EA8;
+  --blue: #4888ee;
+  --red: #d80027;
+  --gray: #a19ea8;
+  --dark-gray: #736e7d;
+  --dark-gray-rgb: 115, 110, 125;
```
Comment (andyesp) [inline src/theme.css:152]: You can include rgb/rgba here. If the color has opacity I'd add it here as a color too.
For example, we can have `dark-gray` and `dark-gray-opacity` or similar.

**Review by andyesp — CHANGES_REQUESTED** [review]

File: src/theme.css (hunk 4 above)
Comment (ncomerci) [reply src/theme.css:152]: Yes I know, but actually the opacity depends on the component that uses the color. That's why I defined it this way, I leave you some examples below:

image
image
image
image

We could agree on a standard with Juani and define an opacity value, WDTY?

Comment (andyesp) [reply src/theme.css:152]: Yes! Let's define a standard opacity value

File: src/components/Section/DetailsSection.css [hunk 5]
```diff
@@ -93,10 +109,27 @@
...
+@media (min-width: 1080px) {
+  .DetailsSection .DetailsSection__Flex {
+    flex-direction: row;
+  }
+
+  .DetailsSection .DetailsSection__Flex--Coauthors span svg {
+    min-width: 10px;
+  }
+
+  /* .DetailsSection .DetailsSection__Flex:nth-child(3) {
```
Comment (andyesp) [inline src/components/Section/DetailsSection.css:132]: Comments!

File: src/hooks/useCoauthoring.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import useAsyncMemo from 'decentraland-gatsby/dist/hooks/useAsyncMemo'
+
+import { Governance } from '../api/Governance'
+import { CoauthorAttributes, CoauthorStatus } from '../entities/Coauthor/types'
+
+function useCoauthoring(coauthor?: string | null, status?: CoauthorStatus) {
```
Comment (andyesp) [inline src/hooks/useCoauthoring.ts:6]: Maybe `useProposalsByCoAuthor` is more specific of what this hook does

File: src/components/Section/ProposalCoAuthorStatus.tsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+  const t = useFormatMessage()
+
+  const updateStatus = async (status: CoauthorStatus) => {
+    await Governance.get().updateCoauthorStatus(proposalId, status)
+    location.reload()
+  }
+
+  const isPending = status === CoauthorStatus.PENDING
+  const pending = isPending ? 'Pending' : ''
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/components/Section/ProposalCoAuthorStatus.tsx:69]: I use it later to conditionally style the component
image

File: src/entities/Coauthor/routes.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  const result = await CoauthorModel.update<CoauthorAttributes>({ status: data.status }, conditions)
+
+  if (result.rowCount === 1) {
+    return {
+      ...conditions,
+      status: data.status,
+    }
+  }
+
+  return null
```
Comment (1emu) [inline src/entities/Coauthor/routes.ts:47]: This return null is suspicious... does this method allow for an update of more than one record?
If result.rowCount !== 1 is an unexpected error, we should throw.

File: src/intl/en.json [hunk 9]
```diff
@@ -484,6 +496,10 @@
       },
       "details_label": "Details",
       "details_user_label": "Created by",
+      "details_coauthors_label": "Co-authors",
+      "details_coauthor_accepted_helper": "This user has accepted the co-author request",
+      "details_coauthor_rejected_helper": "This user has declined the co-author request",
+      "details_coauthor_pending_helper": "This user has the co-author status as pending",
```
Comment (1emu) [inline src/intl/en.json:502]: Wording suggestion: "This user has a pending co-author request"

**Review by 1emu — APPROVED** [review]
LGTM! Left some minor comments 
🥳

File: src/entities/Proposal/routes.ts [hunk 10]
```diff
@@ -377,6 +390,13 @@ export async function createProposal(
   const start = Time.utc().set('seconds', 0)
   const end = data.finish_at
   const proposal_url = proposalUrl({ id })
+  let coAuthors: string[] | null = null
```
Comment (andyesp) [inline src/entities/Proposal/routes.ts:393]: I think you can write this:
```
const coAuthors = (data.configuration && data.configuration.coAuthors) ? data.configuration.coAuthors : null
```

File: src/entities/Proposal/routes.ts [hunk 11]
```diff
@@ -377,6 +390,13 @@ export async function createProposal(
...
+
+  if (data.configuration && data.configuration.coAuthors) {
+    coAuthors = data.configuration.coAuthors
+    delete data.configuration.coAuthors
```
Comment (andyesp) [inline src/entities/Proposal/routes.ts:397]: Is it necessary to delete this from configuration?

File: src/pages/index.tsx [hunk 12]
```diff
@@ -206,6 +212,7 @@ export default function IndexPage() {
                       <ProposalItem
                         key={proposal.id}
                         proposal={proposal}
+                        coauthorRequest={!!pendingCoauthorRequests.find((req) => req.proposal_id === proposal.id)}
```
Comment (andyesp) [inline src/pages/index.tsx:215]: If you aren't passing the entire request object as prop and `coauthorRequest` is a boolean I'd call this `hasCoauthorRequest`

File: src/entities/Proposal/routes.ts (hunk 11 above)
Comment (ncomerci) [reply src/entities/Proposal/routes.ts:397]: I don't see why it should be stored in configuration as well if we have a specific table for this. WDYT?

File: src/api/Governance.ts [hunk 13]
```diff
@@ -285,4 +288,25 @@ export class Governance extends GovernanceAPI {
...
+    const newsSatus = await this.fetch<ApiResponse<CoauthorAttributes>>(
```
Comment (andyesp) [inline src/api/Governance.ts:305]: I think this variable name should be `newStatus` 😛

**Review by andyesp — APPROVED** [review]

**Review by 1emu — APPROVED** [review]

**Review by andyesp — APPROVED** [review]

**Pull request merged**